### DISCUSSION TITLE
xPub wallet should be a priority

### DISCUSSION THREAD

**Discussion by GiverofMemory** [dogecoin/dogecoin#2831] [General]
So in light of the censorship via crowdfunding platforms and also the new $600 IRS reporting requirement for payment processors, we should focus on crypto fundraising.  In terms of what is needed from developers is an xPub, preferably a dogecoin fork of Electrum wallet like Bitcoin cash has done with [Electron Cash](https://electroncash.org/) and [source code](https://github.com/Electron-Cash/Electron-Cash).  

What an xPub does is provide you with a public key that can be used to securely generate endless private and public key combo's.  This can be used especially for merchants so they can generate a new public key for every purchase automatically. 

Also if we fork Electrum in order to have a wallet with this functionality, it can easily be integrated into powerful direct payment processors like [BitcartCC](https://bitcartcc.com/).  I have talked to the devs of this excellent project and they said that if the Dogecoin Devs forked electrum, they would integrate dogecoin payment functionality to the platform.

Dogecoin is uncensorable money and the world needs this very much right now.

**Comment by MrNaif2018** [comment]
Hello! BitcartCC dev here. If dogecoin maintainers decide to create electrum fork, I can guide you initially.
For example here's an example how I ported electrum to bsty coin:
https://github.com/GlobalBoost/electrum/commit/6d7c7a1cb5c384377b24391bcbc7b797462b980a

Our policy for adding a new coin is basically that, once in a while when we do scheduled electrum's upgrade (like this one https://github.com/bitcartcc/bitcart/pull/234), you would have to upgrade your fork to some upstream commit we select (usually at releases, but sometimes right after my patches to electrum)
After initial implementation is created, just upgrading once in a while has almost no maintenance burden and is pretty quick

Plus as a bonus you might consider adding more features to it or just more than just initial re-branding, then dogecoin users get an amazing non-custodial and secure wallet they can use

At BitcartCC we pin all electrums to a tested version, same for all coins, so we have the same API and compatibility layers everywhere. When upgrading we check if there were breaking changes and adjust the code, and the whole payment processing is based on that core code based on electrum

**Comment by patricklodder** [comment]
Hi! The [not-bitcoin-sv fork of ElectrumX](https://github.com/spesmilo/electrumx) that the forks you link are branched off of, supports Dogecoin out of the box, [extended pubkeys per BIP32](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/chainparams.cpp#L170) have been specified since 1.7.0 in 2014 and [Dogecoin was one of the initially listed coins in SLIP44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md). Could you please elaborate on what is needed precisely?

**Reply by MrNaif2018** [reply]
Hi! ElectrumX server implementation is just one step, and it's already done, that's great. But in order to be secure we need all that SPV technology client-side. So basically a fork of electrum client wallet should be created:
https://github.com/spesmilo/electrum
In the comment above I've sent a commit showing how I did it for another currency. You can re-brand in any way you like, change settings to make it sync (for example if another hash function is used, or if another address prefix is used) in file networks.py, run a few electrumx servers and encourage people to run them too, and add initial trusted servers to servers.json.
When it's all done what you get is an amazing and secure wallet other users can use (you can build binaries via script the repo has), and when it's done adding dogecoin is done in BitcartCC in less than an hour, as electrum is ready (we import it as a python library)

**Reply by patricklodder** [reply]
Alright. Thank you for elaborating. If someone needs help getting this done I can make time for supporting.

**Reply by GiverofMemory** [reply]
Wow thank you Naif for that explanation!  I will take a look at this and see if I can do some of these steps but will probably have questions.

**Reply by GiverofMemory** [reply]
I am letting you all know I am naming it ElectroDoge so let me know if you think we should name it something different before I change all the names.  Also I picked up electrodoge.com we can use for a user friendly way for people to get it (I am fine transferring the domain to dogecoin dev ownership, just let me know).  But regardless I am fine naming it something different just let me know.  Here is the repo I am working on and I suppose dogecoin can fork it when it is done or I can just give devs access to my repo, either way I am fine with. https://github.com/GiverofMemory/electrodoge

**Comment by GiverofMemory** [comment]
Question: What is the equivalent of this bitcoin info on dogecoin?  If you don't know off the top of your head, what files should I be looking in?  Thanks @patricklodder 

    WIF_PREFIX = 0x80
    ADDRTYPE_P2PKH = 0
    ADDRTYPE_P2SH = 5
    SEGWIT_HRP = "bc"
    GENESIS = "000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f"

Context: https://github.com/GlobalBoost/electrum/blob/master/electrum/constants.py#L73

**Reply by patricklodder** [reply]
See: https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/chainparams.cpp#L167-L171 for most of it.

`GENESIS` is the hash of block 0, can be found in mainnet checkpoints in that same file

`SEGWIT_HRP` is the bech32 HRP and this is set to `doge` per 1.21 decisions made: https://github.com/dogecoin/dogecoin/blob/20b3f31035937e4d1e4b0962daea786ab34d13ec/src/chainparams.cpp#L141 - do note that segwit is not implemented.

**Reply by GiverofMemory** [reply]
Ok I appreciate it.

**Comment by GiverofMemory** [comment]
@MrNaif2018 when you have a sec, could you help me decode what is needed for these lines?

This is what we have:
https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/chainparams.cpp#L167-L171

How would that inform these settings, what should I change them to?

    WIF_PREFIX = 0x80
    ADDRTYPE_P2PKH = 0
    ADDRTYPE_P2SH = 5

ref: https://github.com/GlobalBoost/electrum/blob/master/electrum/constants.py#L73

Also I forked the latest electrum instead of the version you did, I tried to make all the same changes.  Will that be a problem?  Also in terms of the IP addresses for servers, how is that done exactly, basically it runs like a standard blockchain node and you list your public IP?  Is there an easy way to make that private, perhaps by routing it through a domain name?

**Reply by MrNaif2018** [reply]
Well, it's simple mostly.
ADDRTYPE_P2PKH should be 30, ADDRTYPE_P2SH should be 22 (see, it's second argument in vector init)
And WIF_PREFIX is 0x9e (hex value of 158)

Check if doge has segwit and also submit SEGWIT_HRP

set GENESIS to the actual genesis block

Also check xprv and other headers, were they modified? From what I see, yes (I guess you can just concatenate those)

Forking the latest electrum is fine. Note that bitcart right now doesn't use the latest-latest electrum release, check pinned version here: https://github.com/bitcartcc/bitcart/blob/master/requirements/daemons/btc.txt

But in a few days when I'm more free we will do a scheduled electrums upgrade anyway.
But if I'm not mistaken, latest electrum releases now create segwit-only wallets, so check if that's desired

In terms of IP for servers, no, you need to run a full node, then also electrumx indexer server, and point electrum to electrumx: https://github.com/spesmilo/electrumx. The more nodes the better, at least more than 1 node would be good. Domain name works too. But I guess you know that domain name translates into the same IP anyway?

**Reply by GiverofMemory** [reply]
Thanks for the help.  I am not a coder so hard for me to get a handle on these little intricacies, I might have more questions when I get there, probably with the xprv part.

Dogecoin does not have segwit and I don't think it is planning to but I need to check out current 1.21 directions.

Darn on the segwit only wallets with current versions!  I will probably have to start from square one again after many hours of work...

Thanks again.

**Reply by MrNaif2018** [reply]
Well, to be fair they didn't remove legacy addresses support, not at all. Just that in the UI it'll no longer display an option whether to create a legacy or a segwit wallet. When restoring legacy wallet it all works

**Reply by GiverofMemory** [reply]
Ok, thanks.  Well I want the most streamlined version for wallet users.  I just checked and Dogecoin has no plans on activating segwit, so what would be the very best version of electrum to fork at this point?  I will use that.  Are you activating segwit and lightning on globalboost?  If not, I can just follow your lead on maintenance.

**Reply by MrNaif2018** [reply]
Well, I activated both segwit and LN on globalboost in electrum but not sure if it was supported :D You can use the most recent copy, just modify the UI to create legacy wallets. For bitcart to function it depends on some CLI/wallet commands you would probably never need to modify.
https://github.com/spesmilo/electrum/pull/6806 you might need to revert/adapt this

**Reply by GiverofMemory** [reply]
Ok, cool.  Well does it sound like a good idea that I just take a snapshot 1 commit before the default to segwit?  That would be this: https://github.com/spesmilo/electrum/tree/94708f63e4f237f5f4c143137de797f9c0ccc08e

I am hesitant to just revert that one commit and use the wallet in it's current version because I am worried things would be broken or would break in the future (and not being a coder I want something easy to maintain).

Is there any risk of basically adopting an old version, and just merging in any specific security updates only in the future?  Sometimes I have had a wallet (like electron cash SLP) that doesn't "find" the coins anymore and I had to take my seed to a new wallet.  How do I make sure this wallet never stops working?

Thanks again.

**Reply by MrNaif2018** [reply]
Well it depends on how you want to maintain it, but in bitcart we have a requirement where we do electrums upgrade from time to time, so each electrum has the same base version/commit. Best is to take last version, apply blockchain changes and just revert one commit via git revert maybe or somewhat manually

**Reply by GiverofMemory** [reply]
Thanks, so I (manually) reverted the commit and I was already working on the latest version of the code so I don't have to start over at least.  I am mostly done now, just stuck on what to change the XPRV headers to.  I did a search of dogecoin code for "xprv" and found nothing, how should I modify constants.py XPRV headers section for electrum for dogecoin? I don't need to change the testnet section correct?  I appreciate the help.

**Reply by MrNaif2018** [reply]
Hi! Just wanted to tell that we've upgraded electrums to this commit now: spesmilo/electrum@28fe345b0bcf708c06abce8382d00de30d210b75

As for the xprv/xpub, it is basically

0x
and then write all numbers from the arrays
example here: https://github.com/GlobalBoost/GlobalBoost-Y/blob/master/src/chainparams.cpp#L140-L141
see how it translates to electrum: https://github.com/GlobalBoost/electrum/blob/master/electrum/constants.py#L87

**Reply by GiverofMemory** [reply]
Thank you for that info, and thanks so much for the update!  That will help me make sure my version is compatible and everything.  All icons made (took longer than expected), double checking all my files against your original commit (found some errors I made so far), and then will check against your new update.  Hopefully the end is in sight, another month maybe lol.

*Edit: looks like I should probably change checkpoints.json.  Are these electrum checkpoints from running a node or can I find these in dogecoin code?  And if they are electrum checkpoints, how should I approach getting these?

**Comment by daank-c** [comment]
> What an xPub does is provide you with a public key that can be used to securely generate endless private and public key combo's. This can be used especially for merchants so they can generate a new public key for every purchase automatically.

Correct me if I'm wrong,

but I think this type of system is what Tesla uses for their online mug sales (paid in Doge). 
It seems like it's necessary, or at least very useful!

**I just grabbed this screenshot from their site.

![Screenshot 2022-07-05 9 02 07 PM](https://user-images.githubusercontent.com/102710924/177438077-263e087e-f57b-4997-a106-e1e66f28f359.png)

**Reply by GiverofMemory** [reply]
Yes they must be using some sort of xPub tech.  Very critical for merchants so hopefully this wallet can increase merchant adoption!

**Comment by CryptoUnchained** [comment]
Agreed and thanks a million as I have an old lngerhans wallet this has fast forwarded update/upgrade and account balance recovery for. This would’ve taken me years to figure out on my own Sent using the mobile mail appOn 7/31/22 at 18:47, GiverofMemory wrote:
                
            
            
                
                    From: "GiverofMemory" ***@***.***>Date: July 31, 2022To: "dogecoin/dogecoin" ***@***.***>Cc: "Subscribed" ***@***.***>Subject: Re: [dogecoin/dogecoin] xPub wallet should be a priority (Discussion #2831)
                
                
Thank you for that info, and thanks so much for the update!  That will help me make sure my version is compatible and everything.  All icons made (took longer than expected), double checking all my files against your original commit (found some errors I made so far), and then will check against your new update.  Hopefully the end is in sight, another month maybe lol.

**Comment by GiverofMemory** [comment]
As an update I got the initial commit loaded up, and merged all electrum updates since the time I started https://github.com/GiverofMemory/electrodoge.  So I am happy about that.  It still isn't "ready" yet as we don't have any servers running.

Here are my notes on future direction and stuff:

typo wallet.py line 396

servers_testnet.json should be ok uses localhost

servers.json needs IP change when a server is up.

checkpoints.json needs update

not sure if android_electrum_icon_foreground.png needs to be updated.

investigate <<<<<HEAD and ====== in code https://github.com/spesmilo/electrum/compare/master...GiverofMemory:electrodoge:main

**Reply by MrNaif2018** [reply]
Hi! I see it has server in it, is it already running or is it a dummy one? Anyways when you're done let me know, will help rebasing and will integrate it

**Reply by MrNaif2018** [reply]
Just pushed rebasing changes. Server seems to work, no? Also please ideally open issues on electrodoge repo + join one of BitcartCC communities to coordinate the changes

**Reply by patricklodder** [reply]
I have limited time, but this is actually starting to get more and more important because of other initiatives I see people working on in the field. How can I help?

**Reply by MrNaif2018** [reply]
Oh, well, sorry, missed the message! The best way to get this running is to launch a network of ElectrumX servers for Dogecoin. The more the better obviously
https://github.com/spesmilo/electrumx
Current electrodoge client seems to work just fine right now, but it somehow? connects to a BSTY server as there are no DOGE servers yet
Let me know if at least one is running to re-verify whether client works

**Reply by patricklodder** [reply]
👍 this I can do. Will look at this somewhere this week.

**Reply by patricklodder** [reply]
It's up.

tcp: `mainnet.electrum.plddr.eu:50001`
ws: `mainnet.electrum.plddr.eu:50003`

I'll switch to TLS once there is at least 1 production application that needs it. 

Note for others installing this: I've seen `electrumx_server` bursting to 6GB RAM at startup, so for a real production node (with local dogecoind, because shouldn't expose RPC other than localhost) I would suggest one would minimally need:
- RAM: 12GB
- CPU cores: 4
- Disk: 120GB SSD or NVMe

**Reply by GiverofMemory** [reply]
Awesome thanks Patrick :).  So there still may need to be some work done in servers.json checkpoints.json  both of which I am unsure about.  But I will try to keep this on the chaintip of electrum updates and would love to make more maintainers as well so you guys can push any changes you need.

>but it somehow? connects to a BSTY server as there are no DOGE servers yet

Ya thats my bad, I will need to add these servers from patrick...lemme see if I can figure it out from the info he provided.

Ok so updated [servers.json](https://github.com/GiverofMemory/electrodoge/blob/main/electrum/servers.json) check it out and see if I did it right ;)

**Reply by patricklodder** [reply]
I'm not sure about the version. It says `'v1.4.2'` when I call `server.peers.subscribe`

**Reply by MrNaif2018** [reply]
It seems to be kind of stuck during the time auxpow was activated:
```
W | NotificationSession | MemoryError('dropping message over 1,000,000 bytes and re-synchronizing')
I/i | interface.[mainnet.electrum.plddr.eu:50001] | disconnecting due to RequestTimedOut("request timed out: ('blockchain.block.headers', [370944, 2016]) (id: 43)")
```

Does the server have enough RAM?

EDIT: modifying client-side constant, it now tells this:
`
disconnecting due to RequestTimedOut("request timed out: ('blockchain.scripthash.subscribe', ['059cee473b5beee8b533eed394f5d69d5e74bc0765ba51bbb0eb194cabce8bbc']) (id: 37)")`

Is the data transferred so large? (timeout is 30 seconds)

**Reply by MrNaif2018** [reply]
Ignore my previous message, it was just client-side constants mostly. 

Well, I tried my best, but it is stuck near auxpow block, giving negative target somehow, probably due to wrong calculation. I would need your help, see https://github.com/GiverofMemory/electrodoge/pull/3

**Reply by GiverofMemory** [reply]
Not sure if checkpoints.py is a problem, would those be the same as the main blockchain checkpoints or electrum server checkpoints? (which presumably we wouldn't have any yet)

**Reply by patricklodder** [reply]
> Does the server have enough RAM?

It should have (it has about 4Gb available), but I'm seeing some of these:

Sep 20 08:50:59 ld6 electrumx_server[1718]: INFO:AuxPoWElectrumX:[244] changing task concurrency from 10 to 6
Sep 20 08:51:02 ld6 electrumx_server[1718]: INFO:AuxPoWElectrumX:[244] disconnected whilst throttled

I will try to troubleshoot/tune on my end because I've kept all the default limits.

I will at some point migrate this to either k8s or lxd infra on an even bigger box so that I can integrate it with the whole infra-as-code thing I have going, but I honestly didn't want to do that until I have figured out (a) monitoring and (b) long-term security, because of that odd mention of client side permanent cert pinning (which does not fit any security model I like because how would I do rollover or implement crl?!?)

> Ignore my previous message, it was just client-side constants mostly.

Can you share which constants you needed to change, so that I can attempt to match them to server-side configuration and improve it?

>  it is stuck near auxpow block, giving negative target somehow, probably due to wrong calculation

I remember a discussion about the NMC implementation a couple months ago, let me try to figure out what was the final result there, because there were issues with the old Dogecoin auxpow code. The code on the server side works though (some 18 months ago I helped the troubleshooting that resulted in this: https://github.com/spesmilo/electrumx/commit/6b077db6790d599817b6467d047220e8b497a125) so that may be a good source?

> Not sure if checkpoints.py is a problem

I saw some merge artifacts in there yesterday. I couldn't find documentation on the format, will dig deeper.

**Reply by GiverofMemory** [reply]
Checkpoints.json I meant.  All the current checkpoints are probably garbage from what I remember, if I was unsure about anything I just copied what globalboost had so ya somehow we need to generate our own.

**Reply by MrNaif2018** [reply]
Note: `checkpoints.json` is not a problem for now. Someone needs to first sync from genesis up to the current block and then export them via `network.export_checkpoints` or like so. 
I tried doing this, but my sync is stuck around auxpow activation with many different errors, so looks like target and bits calculation used in electrum/btc is not the same in doge, but I don't know how to implement it. It's mostly all in blockchain.py `get_target` from what I see

>  some 18 months ago I helped the troubleshooting that resulted in this: https://github.com/spesmilo/electrumx/commit/6b077db6790d599817b6467d047220e8b497a125) so that may be a good source?

Well, I don't see target calculation here, it is probably not even needed for electrumx server, but I'll re-check (Yes, it just gives bits it receives from the full node)

> Can you share which constants you needed to change, so that I can attempt to match them to server-side configuration and improve it?

Sure, @patricklodder see this issue https://github.com/spesmilo/electrum/issues/4315

It's  MAX_INCOMING_MSG_SIZE = 1_000_000  in electrum/interface.py
Server-side setting is https://electrumx-spesmilo.readthedocs.io/en/latest/environment.html#envvar-MAX_SEND

But the cool thing is, defaults for auxpow are already adjusted. So it was client-side only

**Reply by patricklodder** [reply]
> It's mostly all in blockchain.py get_target from what I see

Ok, this is [digishield](https://dgbwiki.com/index.php?title=DigiShield) (per block retargeting w/ amplitude.) We'll have to port the code because it's all we have - there is no real spec outside of the code. I'll make a start with that because that will refresh my memory on it and help me write the spec later, as that's an [open task for this repo](https://github.com/dogecoin/dogecoin/issues/3002)

> But the cool thing is, defaults for auxpow are already adjusted. So it was client-side only

Alright. I've also read up on the concurrency thing and I guess I'll need to just trial and error values there, because what it looks like is that `COST_SOFT_LIMIT` is causing the server to scale its capacity down automagically, so I'll just have to raise these values until it breaks. Will test with that later.

**Reply by MrNaif2018** [reply]
You can also set it to 0 to remove limits if you wish so

**Reply by patricklodder** [reply]
> You can also set it to 0 to remove limits if you wish so

interesting - will do that and see how it goes.

**Reply by GiverofMemory** [reply]
So it keeps failing to easily merge because they are updating checkpoints on the spiesmilo version.  Can I use gitignore to ignore their updates to checkpoints?

**Reply by patricklodder** [reply]
I wouldn't use gitignore for that but instead of doing `git merge` on upstream changes, look at the commits and `git cherry-pick` the ones you need (and exclude those that change checkpoints)

**Reply by MrNaif2018** [reply]
I wouldn't exclude any commits, it could get broken easily (also electrum often does follow-up fixes as new commits)
The easiest way is to just run merge and resolve conflicts, there aren't much. In vscode for example it's one click to discard their changes 

As for the verification code, @patricklodder did you get a chance to look at it?

**Reply by MrNaif2018** [reply]
Hi @patricklodder! Did you get a chance to look at the digishield verification code?

P.S. @GiverofMemory I've seen your reddit thread, no, bitcart can't add dogecoin without that verification code

**Reply by GiverofMemory** [reply]
> I wouldn't exclude any commits, it could get broken easily (also electrum often does follow-up fixes as new commits) The easiest way is to just run merge and resolve conflicts, there aren't much. In vscode for example it's one click to discard their changes
> 
Thanks it hasn't been too bad I have been able to just discard my commits to update, then remerge my changes without issue.

**Comment by CryptoUnchained** [comment]
Once Doge radio is live with new wallet and card capabilities should rival XRP and ADA in that department IMO Sent using the mobile mail appOn 7/31/22 at 18:45, GiverofMemory wrote:
                
            
            
                
                    From: "GiverofMemory" ***@***.***>Date: July 31, 2022To: "dogecoin/dogecoin" ***@***.***>Cc: "Subscribed" ***@***.***>Subject: Re: [dogecoin/dogecoin] xPub wallet should be a priority (Discussion #2831)
                
                
Yes they must be using some sort of xPub tech.  Very critical for merchants so hopefully this wallet can increase merchant adoption!

**Comment by vasuxdev** [comment]
Is there any latest release with all fixes?

**Reply by GiverofMemory** [reply]
I wouldn't worry about the release, I haven't made binaries yet, it is just a zip file of the same thing you can download the repo.

**Comment by EvensD** [comment]
HOW CAN I DO TO MAKE THE DOGECOIN
